api/desecapi/tests/test_domains.py

```diff
@@ -163,6 +164,14 @@ def test_create_under_public_suffix_with_private_parent(self):
             response = self.client.post(self.reverse('v1:domain-list'), {'name': name})
             self.assertStatus(response, status.HTTP_201_CREATED)
 
+    def test_create_domain_under_unsupported_public_suffix_rule(self):
+        # Show lenience if the PSL library produces an UnsupportedRule exception
+        name = 'unsupported.wildcard.test'
+        psl_cm = self.get_psl_context_manager(UnsupportedRule)
```

Comment:
If I understand correctly, this passes an `Exception` type object to the `public_suffix` argument in `_mock_get_public_suffix`. That this actually works is quite surprising, consider renaming the argument of that function

Reply:
The code passes a class (derived from `Exception`) to `get_psl_context_manager`. The class is callable (calling it produces the instance). `get_psl_context_manager` checks for this case, and if the thing at hand is callable, it does not pass it on to `_mock_get_public_suffix`, but instead uses it directly to set the `side_effect` attribute of the mocked function.

Side effects in turn are one of the neat features of `unittest.mock`. You can set a `side_effect` attribute on your mocked objects/methods/... which takes effect (haha) when the mock is called. They have logic that if the side effect is an `Exception` instance, it is raised and not called. That's actually pretty cool, because you can quickly simulate various error scenarios. https://docs.python.org/3/library/unittest.mock.html#quick-guide

One could pass any other callable (most useful would be a function) to the context manager. The function then has to either produce a return value for the mocked method, or raise an exception. I was amazed how flexible the mock library is.

Because of the flexibility, it's hard to find a good name for that argument, so I made a bad choice :) I changed it to `side_effect_parameter`.